Intersect.Server.Core/Entities/Player.cs

```diff
@@ -1275,51 +1276,100 @@ public void SetLevel(int level, bool resetExperience = false)
         PacketSender.SendExperience(this);
     }
 
-    public void LevelUp(bool resetExperience = true, int levels = 1)
+    public void AddLevels(int levels = 1, bool resetExperience = true)
     {
-        var messages = new List<string>();
-        if (Level < Options.Instance.Player.MaxLevel)
+        if (levels > 0)
         {
-            for (var i = 0; i < levels; i++)
+            while (levels > 0)
             {
-                SetLevel(Level + 1, resetExperience);
-
-                //Let's pull up class - leveling info
-                var classDescriptor = ClassBase.Get(ClassId);
-                if (classDescriptor?.Spells == null)
+                if (Level >= Options.Instance.Player.MaxLevel)
                 {
-                    continue;
+                    break;
                 }
 
-                foreach (var spell in classDescriptor.Spells)
+                SetLevel(Level + 1, resetExperience);
+                StartCommonEventsWithTrigger(CommonEventTrigger.LevelUp);
+
+                PacketSender.SendChatMsg(
+                    this,
+                    Strings.Player.LevelUp.ToString(Level),
+                    ChatMessageType.Experience,
+                    CustomColors.Combat.LevelUp,
+                    Name
+                );
+
+                if (ClassBase.TryGet(ClassId, out var classDescriptor) && classDescriptor?.Spells != default)
```

Comment:
Put `ClassBase? classDescriptor = null;` before `if (levels > 0)`

Here, change the `if` to `if ((classDescriptor?.Id == ClassId || ClassBase.TryGet(ClassId, out classDescriptor)) && classDescriptor?.Spells != default)`